langchain/src/stores/message/firestore.ts

```diff
@@ -0,0 +1,160 @@
+import * as firebaseAdmin from "firebase-admin";
+import { getApps, initializeApp } from "firebase-admin/app";
+
+import {
+  getFirestore,
+  DocumentData,
+  Firestore,
+  DocumentReference,
+  FieldValue,
+} from "firebase-admin/firestore";
+import {
+  StoredMessage,
+  BaseMessage,
+  BaseListChatMessageHistory,
+} from "../../schema/index.js";
+import {
+  mapChatMessagesToStoredMessages,
+  mapStoredMessagesToChatMessages,
+} from "./utils.js";
+
+export interface FirestoreDBChatMessageHistory {
+  collectionName: string;
+  sessionId: string;
+  userId: string;
+  appIdx?: number;
+  config?: firebaseAdmin.AppOptions;
+}
+export class FirestoreChatMessageHistory extends BaseListChatMessageHistory {
+  lc_namespace = ["langchain", "stores", "message", "firestore"];
+
+  private collectionName: string;
+
+  private sessionId: string;
+
+  private userId: string;
+
+  private appIdx: number;
+
+  private config: firebaseAdmin.AppOptions;
+
+  private firestoreClient: Firestore;
+
+  private document: DocumentReference<DocumentData> | null;
+
+  private messages: BaseMessage[];
+
+  constructor({
+    collectionName,
+    sessionId,
+    userId,
+    appIdx = 0,
```

Comment:
What is `appIdx`? Basically `conversationId`?

Reply:
appIdx is index of app configuration.  firebase functions supports multiple app configurations. user can manually pass a config or if firebase functions are pre-initialized, user can pass appIdx - the index of the app 

https://firebase.google.com/docs/projects/multiprojects#web_1